core/CustomMethods.php

```diff
@@ -0,0 +1,252 @@
+<?php
+
+namespace SilverStripe\Framework\Core;
+
+use Exception;
+
+/**
+ * Allows an object to declare a set of custom methods
+ */
+trait CustomMethods {
+
+    /**
+     * Custom method sources
+     *
+     * @var array
+     */
+	protected static $extra_methods = array();
+
+	/**
+	 * Name of methods to invoke by defineMethods for this instance
+	 *
+	 * @var array
+	 */
+	protected $extra_method_registers = array();
+
+    /**
+     * Non-custom methods
+     *
+     * @var array
+     */
+    protected static $built_in_methods = array();
+
+    /**
+     * Attempts to locate and call a method dynamically added to a class at runtime if a default cannot be located
+     *
+     * You can add extra methods to a class using {@link Extensions}, {@link Object::createMethod()} or
+     * {@link Object::addWrapperMethod()}
+     *
+     * @param string $method
+     * @param array $arguments
+     * @return mixed
+     * @throws \Exception
+     */
+	public function __call($method, $arguments) {
+		// If the method cache was cleared by an an Object::add_extension() / Object::remove_extension()
+		// call, then we should rebuild it.
+        $class = get_class($this);
+		if(empty(self::$extra_methods[$class])) {
+			$this->defineMethods();
+		}
+
+		$method = strtolower($method);
+
+		if(isset(self::$extra_methods[$class][$method])) {
+			$config = self::$extra_methods[$class][$method];
+
+			switch(true) {
+				case isset($config['property']) :
+					$obj = $config['index'] !== null ?
+						$this->{$config['property']}[$config['index']] :
+						$this->{$config['property']};
+
+					if($obj) {
+						if(!empty($config['callSetOwnerFirst'])) $obj->setOwner($this);
+						$retVal = call_user_func_array(array($obj, $method), $arguments);
+						if(!empty($config['callSetOwnerFirst'])) $obj->clearOwner();
+						return $retVal;
+					}
+
+					if(!empty($this->destroyed)) {
+						throw new \Exception (
+							"Object->__call(): attempt to call $method on a destroyed $class object"
+						);
+					} else {
+						throw new \Exception (
+							"Object->__call(): $class cannot pass control to $config[property]($config[index])."
+								. ' Perhaps this object was mistakenly destroyed?'
+						);
+					}
+
+				case isset($config['wrap']) :
+					array_unshift($arguments, $config['method']);
+					return call_user_func_array(array($this, $config['wrap']), $arguments);
+
+				case isset($config['function']) :
+					return $config['function']($this, $arguments);
+
+				default :
+					throw new \Exception (
+						"Object->__call(): extra method $method is invalid on $class:"
+							. var_export($config, true)
+					);
+			}
+		} else {
+			// Please do not change the exception code number below.
+			$class = get_class($this);
+			throw new \Exception("Object->__call(): the method '$method' does not exist on '$class'", 2175);
```

Comment:
It'll get fixed later when I fix virtualpage. See https://github.com/silverstripe/silverstripe-cms/issues/1386